### PR TITLE
revert to old Switch-style, keeping unmountScenes.  needs review

### PR DESCRIPTION
**Pull request by joenoon**
refs https://github.com/aksonov/react-native-router-flux/issues/976
refs https://github.com/aksonov/react-native-router-flux/issues/952

### CODE REVIEW

**Comment by joenoon** [comment]
Can someone verify unmountScenes works?  It may need to be moved before the `Actions[selectedKey]();`

File: src/Switch.js [hunk 1]
```diff
@@ -1,59 +1,71 @@
...
+        ...navigationState,
+        children: [navState.children[navigationState.index]],
+        index: 0,
+      };
+    }
+
+    this.setState({ navigationState });
   }
-  let navigationState = index !== navState.index ? { ...navState, index } : navState;
-
```
Comment (zxcpoiu) [inline src/Switch.js]: I guess ` if (props.unmountScenes)` should wrapped in line 44
if selected scene != current scene ( scene changed), than we should unmount scene

Comment (zxcpoiu) [inline src/Switch.js]: or
use
`let navigationState = index !== navState.index ? { ...navState, index } : navState;`
to restore `navigationState`  back to `navState`

**Comment by zxcpoiu** [comment]
I had tested in the following structure:

```jsx

// --- this got infinite loop
<Scene key="switch" initial={true} component={Switch} tabs={true} unmountScenes={true} selector={switchSelector}>
    <Scene key="login" initial={true}/>
    <Scene key="home"/>
</Scene>
```

and

```jsx
// --- this would not switch even selector returns 'home'

<Scene key="switch" initial={true} component={Switch} tabs={true} unmountScenes={true} selector={switchSelector}>
    <Scene key="login"/>
    <Scene key="home" initial={true}/>
</Scene>
```

**Comment by joenoon** [comment]
@zxcpoiu thanks for checking! Just made a change to only do the unmountScenes logic when a switch will actually occur.

@ms88privat could you also verify?

**Comment by sarovin** [comment]
@joenoon I can test it tomorrow

**Comment by zxcpoiu** [comment]
I still got infinite loop in the first case with new commits.

It feels like a little weird to me, just want to make sure, is the following descriptions correct or not:

at the first case:
```jsx
// --- this got infinite loop
<Scene key="switch" initial={true} component={Switch} tabs={true} unmountScenes={true} selector={switchSelector}>
    <Scene key="login" initial={true}/>
    <Scene key="home"/>
</Scene>
```


* `navState`'s children is `login` and `home`, and initial index is '0' since `login` is the initial component
* that's assume in `selector`, we return 'home', then the selected `navigationState` is `home` and it's index mostly is 0 (or the index of it's initial child.)

so the reason I got infinite loop is because this `children: [navState.children[navigationState.index]],`
we always set navState to it's index-0 child which is `login`, but the selector return 'home' which is the index-1 child of navState and has it's own child indexes cache whcih is 0.

and no matter how, eventually, we've change the navState to selected child's navigationState
( I'm not sure, is this the right flow that we replace navState with child's state?)

**Comment by zxcpoiu** [comment]
I have tried several ways, but with no luck so far.

If someone want to reproduce, here is test case:

```jsx
// --- this got infinite loop if switch selector returns 'home'

<Scene key="modal" component={Modal} >
    <Scene key="root">
        <Scene key="switch" initial={true} component={Switch} tabs={true} unmountScenes={true} selector={switchSelector}>
            <Scene key="login" initial={true} type="replace" />
            <Scene key="home" tabs={true} type="replace">
                <Scene key="tab1" initial={true} />
                <Scene key="tab2" initial={true} />
            </Scene>
        </Scene>
    </Scene>
</Scene>
```

**Comment by sarovin** [comment]
```
@@ -15,6 +15,7 @@ export default class Switch extends Component {
   constructor(props, context) {
     super(props, context);
     this.updateState = this.updateState.bind(this);
+    this.resetHistoryStack = this.resetHistoryStack.bind(this);
     this.state = {};
   }
 
@@ -26,6 +27,23 @@ export default class Switch extends Component {
     this.updateState(props);
   }
 
+  resetHistoryStack(child) {
+    const newChild = child;
+    newChild.index = 0;
+    for (let j = 0; j < child.children.length; j++) {
+      if (child.children[j].initial) {
+        newChild.index = j;
+        if (!child.tabs) {
+          newChild.children = [child.children[j]];
+        }
+      }
+      if (newChild.children[j].children) {
+        newChild.children[j] = this.resetHistoryStack(child.children[j]);
+      }
+    }
+    return newChild;
+  }
+
   updateState(props) {
     const navState = props.navigationState;
 
@@ -43,11 +61,8 @@ export default class Switch extends Component {
 
     if (navigationState.key !== navState.children[navState.index].key) {
       if (props.unmountScenes) {
-        navigationState = {
-          ...navigationState,
-          children: [navState.children[navigationState.index]],
-          index: 0,
-        };
+        navState.children[navState.index] = this.resetHistoryStack(navState.children[navState.index]);
+        navigationState = { ...navState };
       }
       Actions[selectedKey]();
     }
```

**Comment by aksonov** [comment]
@sarovin Maybe i'm wrong, but i believe navigationState should be changed within Reducer, not within Switch. What do you think?

**Comment by sarovin** [comment]
@aksonov you are right but i don't understand how work your Reducer  😕 
I try with reducer as soon as possible

**Comment by aksonov** [comment]
Reducer works in the same way as Redux reducers. You may introduce param for JUMP action and add appropriate logic where JUMP is processed.

**Comment by sarovin** [comment]
#1002

**Comment by aksonov** [comment]
Closing it because of #1002, @joenoon does it work for you?

**Comment by joenoon** [comment]
@aksonov so far so good!  Totally missed all the action in the other PR - nice work @sarovin and everyone else!

**Pull request closed without merging**